staging/src/k8s.io/legacy-cloud-providers/azure/azure_controller_common.go

```diff
@@ -255,7 +257,11 @@ func (c *controllerCommon) DisksAreAttached(diskNames []string, nodeName types.N
 		attached[diskName] = false
 	}
 
-	disks, err := c.getNodeDataDisks(nodeName)
+	// doing stalled read for getNodeDataDisks to ensure we don't call ARM
+	// for every reconcile call. The cache is invalidated after Attach/Detach
+	// disk. So the new entry will be fetched and cached the first time reconcile
+	// loop runs after the Attach/Disk OP which will reflect the latest model.
+	disks, err := c.getNodeDataDisks(nodeName, allowUnsafeRead)
```

Comment:
I think should be `cachedData` @andyzhangx

Reply:
`DisksAreAttached` are attached is called only by the volume controller as part of reconcile operation. This reconcile is done every minute which can account for majority of calls /h. For scenarios we tested - 1 VMSS (100 nodes) with 100 disks attached - Every reconcile call generates 100 ARM calls, which more or less will be ~6k-7k calls per hour. With the cache TTL change to 10 mins instead of 1min, it does come down to ~2600 calls per hour for the scenario which still is quite a lot for a reconcile operation which is only validating if the state hasn't changed. 

All other disk operations except `DisksAreAttached` still use the cached data. The cache is invalidated after disk is Attach/Detach. So the next time the reconcile loop is run, new data will be fetched and we can validate if the state looks fine after the initial Attach/Detach. 

@andyzhangx What are the issues you think we'll encounter by allowing unsafe read only for the reconcile loop in disks?

Reply:
> `DisksAreAttached` is called only by the volume controller as part of reconcile operation. This reconcile is done every minute which can account for majority of calls /h. For scenarios we tested - 1 VMSS (100 nodes) with 100 disks attached - Every reconcile call generates 100 ARM calls, which more or less will be ~6k-7k calls per hour. With the cache TTL change to 10 mins instead of 1min, it does come down to ~2600 calls per hour for the scenario which still is quite a lot for a reconcile operation which is only validating if the state hasn't changed.
> 
> All other disk operations except `DisksAreAttached` still use the cached data. The cache is invalidated after disk is Attach/Detach. So the next time the reconcile loop is run, new data will be fetched and we can validate if the state looks fine after the initial Attach/Detach.
> 
> @andyzhangx What are the issues you think we'll encounter by allowing unsafe read only for the reconcile loop in disks?

thanks for the experiment, I think you are right.